iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/services/Deployments.java

```diff
@@ -0,0 +1,310 @@
+            throw new ExternalDependencyException(message, e);
+        }
+    }
+
+    /**
+     * Does 3 queries to the IoTHub to get the status of the deployment per device.
+     *
+     * @param deploymentId - Deployment id for the deployment to query.
+     * @return Map of deviceId to the {@link DeploymentStatus}.
+     */
+    private Map<String, DeploymentStatus> getDeviceStatuses(String deploymentId) throws IOException {
+        final Set<String> appliedDevices = this.getDevicesInQuery(APPLIED_DEVICES_QUERY, deploymentId);
+        final Set<String> successfulDevices = this.getDevicesInQuery(SUCCESSFUL_DEVICES_QUERY, deploymentId);
+        final Set<String> failedDevices = this.getDevicesInQuery(FAILED_DEVICES_QUERY, deploymentId);
+        Map<String,DeploymentStatus> deviceStatuses = new HashMap<>();
+
+        for (String successfulDevice : successfulDevices) {
+            deviceStatuses.put(successfulDevice, DeploymentStatus.Successful);
+        }
+        for (String failedDevice : failedDevices) {
+            deviceStatuses.put(failedDevice, DeploymentStatus.Failed);
+        }
+        for (String device : appliedDevices) {
+            if (!successfulDevices.contains(device) && !failedDevices.contains(device)) {
+                deviceStatuses.put(device, DeploymentStatus.Pending);
+            }
+        }
+        return deviceStatuses;
+    }
+
+    private void verifyDeploymentArgument(final String argumentName, final String argumentValue)
+            throws InvalidInputException {
+        if (StringUtils.isEmpty(argumentValue)) {
+            throw new InvalidInputException("Invalid input. Must provide a value to " +
+                    argumentName);
+        }
+    }
+
+    private Set<String> getDevicesInQuery(String hubQuery, String deploymentId) throws IOException {
+        final String query = String.format(hubQuery, deploymentId);
+        final SqlQuery sqlQuery = SqlQuery.createSqlQuery("*", SqlQuery.FromType.MODULES, query, null);
+        final Query twinQuery;
+        final Set<String> deviceIds = new HashSet<>();
+
+        try {
+            // TODO: Add pagination
+            twinQuery = deviceTwin.queryTwin(sqlQuery.getQuery());
+        } catch (IotHubException ex) {
+            log.error(String.format("Unable to get devices with query %s in deployment %s", query,
+                    deploymentId), ex);
+            return deviceIds;
+        }
+
+        try {
+            while (twinQuery.hasNext()) {
+                final DeviceTwinDevice device = deviceTwin.getNextDeviceTwin(twinQuery);
+                deviceIds.add(device.getDeviceId());
+            }
+        }
+        catch (Exception ex) {
+            log.error("Error getting status of devices in query " + sqlQuery.toString());
+        }
+        return deviceIds;
+    }
+
+    private static Configuration createEdgeConfiguration(final DeploymentServiceModel deployment) throws InvalidInputException {
```

Comment:
nit: remove static?

Reply:
Fixed all others but didn't remove the static. I prefer static methods if the method is not modifying / using the state of the class. Which leads to a method that can probably be moved to a separate helper class but since this is the only place it is needed I have it here for now. Let me know if you disagree.